java-checks/src/main/java/org/sonar/java/checks/TooManyStatementsPerLine_S00122_Check.java

```diff
@@ -47,91 +74,231 @@
 @ActivatedByDefault
 @SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.READABILITY)
 @SqaleConstantRemediation("1min")
-public class TooManyStatementsPerLine_S00122_Check extends SquidCheck<LexerlessGrammar> implements JavaCheck {
-
-  private static final AstNodeType[] STATEMENTS_KINDS = new ImmutableList.Builder<AstNodeType>()
-    .add(
-      Kind.EMPTY_STATEMENT,
-      Kind.LABELED_STATEMENT,
-      Kind.IF_STATEMENT,
-      Kind.ASSERT_STATEMENT,
-      Kind.SWITCH_STATEMENT,
-      Kind.WHILE_STATEMENT,
-      Kind.DO_STATEMENT,
-      Kind.BREAK_STATEMENT,
-      Kind.CONTINUE_STATEMENT,
-      Kind.RETURN_STATEMENT,
-      Kind.THROW_STATEMENT,
-      Kind.SYNCHRONIZED_STATEMENT,
-      Kind.EXPRESSION_STATEMENT,
-      Kind.FOR_STATEMENT,
-      Kind.FOR_EACH_STATEMENT,
-      Kind.TRY_STATEMENT)
-    .build()
-    .toArray(new AstNodeType[0]);
-
-  private final Multiset<Integer> statementsPerLine = HashMultiset.create();
-  private final Map<Integer, Integer> columnsByLine = new HashMap<Integer, Integer>();
+public class TooManyStatementsPerLine_S00122_Check extends SubscriptionBaseVisitor {
+
+  String path;
 
   @Override
-  public void init() {
-    subscribeTo(STATEMENTS_KINDS);
-    subscribeTo(JavaLexer.VARIABLE_DECLARATORS);
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR, Tree.Kind.STATIC_INITIALIZER);
   }
 
-  public boolean isExcluded(AstNode astNode) {
-    return astNode.is(Kind.BLOCK)
-      || astNode.is(Kind.EMPTY_STATEMENT)
-      || astNode.is(Kind.LABELED_STATEMENT)
-      || astNode.getParent().is(JavaLexer.STATEMENT_EXPRESSION);
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    path = context.getFile().getAbsolutePath();
+    super.scanFile(context);
   }
 
   @Override
-  public void visitFile(AstNode astNode) {
-    statementsPerLine.clear();
-    columnsByLine.clear();
+  public void visitNode(Tree tree) {
+    BlockTree block;
+    if (tree.is(Tree.Kind.STATIC_INITIALIZER)) {
+      block = (BlockTree) tree;
+    } else {
+      block = ((MethodTree) tree).block();
+    }
+    if (block != null) {
+      StatementVisitor visitor = new StatementVisitor();
+      block.accept(visitor);
+      for (Entry<Integer> entry : visitor.statementsPerLine.entrySet()) {
+        int count = entry.getCount();
+        if (count > 1) {
+          addIssue(entry.getElement(), "At most one statement is allowed per line, but " + count + " statements were found on this line.");
+        }
+      }
+    }
   }
 
-  @Override
-  public void visitNode(AstNode statementNode) {
-    if (statementNode.is(JavaLexer.VARIABLE_DECLARATORS) && !statementNode.hasParent(JavaLexer.BLOCK_STATEMENTS)) {
-      return;
+  private class StatementVisitor extends BaseTreeVisitor {
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Should org.sonar.java.checks.TooManyStatementsPerLine_S00122_Check$StatementVisitor be a _static_ inner class?[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ASIC_INNER_SHOULD_BE_STATIC)